plugins/upower.c

```diff
@@ -0,0 +1,276 @@
+/*
+ *
+ *  oFono - Open Source Telephony
+ *
+ *  Copyright (C) 2015 Canonical Ltd.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2 as
+ *  published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <glib.h>
+#include <gdbus.h>
+#include <ofono.h>
+
+#define OFONO_API_SUBJECT_TO_CHANGE
+#include <ofono/plugin.h>
+#include <ofono/modem.h>
+#include <ofono/log.h>
+#include <ofono/emulator.h>
+
+#define UPOWER_SERVICE            "org.freedesktop.UPower"
+#define UPOWER_DEVICE_INTERFACE   UPOWER_SERVICE ".Device"
+#define UPOWER_BASE_BATT_PATH     "/org/freedesktop/UPower/devices/battery_"
+#define UPOWER_TOUCH_BATT_PATH    UPOWER_BASE_BATT_PATH "battery"
```

Comment:
Unfortunately it doesn't appear to be possible to add a match which does this filtering for us, as you can't use wildcards in the match rules...

Also, I checked, and it looks like there's some weird behavior with our version of upower, at least on mako.  The EnumerateDevices method shows the following devices:

 * "/org/freedesktop/UPower/devices/line_power_wireless"
 * "/org/freedesktop/UPower/devices/battery_battery"
 * "/org/freedesktop/UPower/devices/line_power_pm8921_dc"
 * "/org/freedesktop/UPower/devices/line_power_touch"

However when I run dbus-monitor, I actually only see 'PropertiesChanged' emitted for two devices, the second which isn't returned by "EnumerateDevices":

 *   "/org/freedesktop/UPower/devices/battery_battery"
 *  "/org/freedesktop/UPower/devices/DisplayDevice"

Note, the "DisplayDevice" always seems include "Type" in it's array of properties ( and it's set incorrectly to 2 ), whereas the "battery_battery" device doesn't include it.   ;(-

So, I think the best I can do is specify the match rules without "path", and then filter for "battery_*" in properties_changed_cb().